cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -982,6 +1021,407 @@ template <typename OutType, typename ArgType, typename Op>
+    ArrayIterator<Arg1Type> arg1_it(arg1);
+    auto arg2_val = UnboxScalar<Arg2Type>::Unbox(arg2);
+    RETURN_NOT_OK(OutputAdapter<OutType>::Write(ctx, out, [&]() -> OutValue {
+      return Op::template Call<OutValue, Arg0Value, Arg1Value, Arg2Value>(
+          ctx, arg0_val, arg1_it(), arg2_val, &st);
+    }));
+    return st;
+  }
+
+  static Status ScalarScalarScalar(KernelContext* ctx, const Scalar& arg0,
+                                   const Scalar& arg1, const Scalar& arg2, Datum* out) {
+    Status st = Status::OK();
+    if (out->scalar()->is_valid) {
+      auto arg0_val = UnboxScalar<Arg0Type>::Unbox(arg0);
+      auto arg1_val = UnboxScalar<Arg1Type>::Unbox(arg1);
+      auto arg2_val = UnboxScalar<Arg2Type>::Unbox(arg2);
+      BoxScalar<OutType>::Box(
+          Op::template Call<OutValue, Arg0Value, Arg1Value, Arg2Value>(
+              ctx, arg0_val, arg1_val, arg2_val, &st),
+          out->scalar().get());
+    }
+    return st;
+  }
+
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch[0].kind() == Datum::ARRAY) {
+      if (batch[1].kind() == Datum::ARRAY) {
+        if (batch[2].kind() == Datum::ARRAY) {
+          return ArrayArrayArray(ctx, *batch[0].array(), *batch[1].array(),
+                                 *batch[2].array(), out);
+        } else {
+          return ArrayArrayScalar(ctx, *batch[0].array(), *batch[1].array(),
+                                  *batch[2].scalar(), out);
+        }
+      } else {
+        if (batch[2].kind() == Datum::ARRAY) {
+          return ArrayScalarArray(ctx, *batch[0].array(), *batch[1].scalar(),
+                                  *batch[2].array(), out);
+        } else {
+          return ArrayScalarScalar(ctx, *batch[0].array(), *batch[1].scalar(),
+                                   *batch[2].scalar(), out);
+        }
+      }
+    } else {
+      if (batch[1].kind() == Datum::ARRAY) {
+        if (batch[2].kind() == Datum::ARRAY) {
+          return ScalarArrayArray(ctx, *batch[0].scalar(), *batch[1].array(),
+                                  *batch[2].array(), out);
+        } else {
+          return ScalarArrayScalar(ctx, *batch[0].scalar(), *batch[1].array(),
+                                   *batch[2].scalar(), out);
+        }
+      } else {
+        if (batch[2].kind() == Datum::ARRAY) {
+          return ScalarScalarArray(ctx, *batch[0].scalar(), *batch[1].scalar(),
+                                   *batch[2].array(), out);
+        } else {
+          return ScalarScalarScalar(ctx, *batch[0].scalar(), *batch[1].scalar(),
+                                    *batch[2].scalar(), out);
+        }
+      }
+    }
+  }
+};
+
+// An alternative to ScalarTernary that Applies a scalar operation with state on
+// only the value pairs which are not-null in both arrays
+template <typename OutType, typename Arg0Type, typename Arg1Type, typename Arg2Type,
+          typename Op>
+struct ScalarTernaryNotNullStateful {
```

Comment:
It seems that this is not used in this pull request?
Should we remove this or use this (by adding support for decimal types)?

Reply:
Decimal support added.

Reply:
Removed unused code, this seems not needed for decimal support.